### ISSUE TITLE
gsutil does not take into account NO_PROXY environment variable

### ISSUE THREAD

**Issue by bquinart** [GoogleCloudPlatform/gsutil#1086]
Similar as #435 but problems seems in another place in the code.

```
Google Cloud SDK 305.0.0
bq 2.0.58
core 2020.08.07
gsutil 4.52
```
Running on Red Hat 7.2

Environment variables like:
```
$ env | grep -i PROXY
NO_PROXY=.googleapis.com
HTTP_PROXY=http://myproxy:8080
HTTPS_PROXY=https://myproxy:8080
```

```
$ gsutil -D ls
...
gsutil version: 4.52
checksum: 38b524a8455fe2a596d5d666b17b8795 (OK)
boto version: 2.49.0
python version: 3.6.8 (default, Jun 11 2019, 15:15:01) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]
OS: Linux 3.10.0-327.el7.x86_64
multiprocessing available: True
using cloud sdk: True
pass cloud sdk credentials to gsutil: True
config path(s): ...
gsutil path: ...
compiled crcmod: False
installed via package manager: False
editable install: False
...
INFO 0814 09:25:53.948534 base_api.py] Making http GET to https://storage.googleapis.com/storage/v1/b?a...
INFO 0814 09:25:53.949151 base_api.py] Body: (none)
Traceback (most recent call last):
  File ".../google-cloud-sdk/platform/gsutil/gsutil", line 21, in <module>
    gsutil.RunMain()
  File ".../google-cloud-sdk/platform/gsutil/gsutil.py", line 123, in RunMain
    sys.exit(gslib.__main__.main())
  File ".../google-cloud-sdk/platform/gsutil/gslib/__main__.py", line 438, in main
    user_project=user_project)
  File ".../google-cloud-sdk/platform/gsutil/gslib/__main__.py", line 767, in _RunNamedCommandAndHandleExceptions
    _HandleUnknownFailure(e)
  File ".../google-cloud-sdk/platform/gsutil/gslib/__main__.py", line 633, in _RunNamedCommandAndHandleExceptions
    user_project=user_project)
  File ".../google-cloud-sdk/platform/gsutil/gslib/command_runner.py", line 411, in RunNamedCommand
    return_code = command_inst.RunCommand()
  File ".../google-cloud-sdk/platform/gsutil/gslib/commands/ls.py", line 563, in RunCommand
    storage_url.scheme).IterBuckets(bucket_fields=bucket_fields):
  File ".../google-cloud-sdk/platform/gsutil/gslib/wildcard_iterator.py", line 479, in IterBuckets
    for blr in self._ExpandBucketWildcards(bucket_fields=bucket_fields):
  File ".../google-cloud-sdk/platform/gsutil/gslib/wildcard_iterator.py", line 398, in _ExpandBucketWildcards
    provider=self.wildcard_url.scheme):
  File ".../google-cloud-sdk/platform/gsutil/gslib/gcs_json_api.py", line 685, in ListBuckets
    global_params=global_params)
  File ".../google-cloud-sdk/platform/gsutil/gslib/third_party/storage_apitools/storage_v1_client.py", line 363, in List
    config, request, global_params=global_params)
  File ".../google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/base_api.py", line 729, in _RunMethod
    http, http_request, **opts)
  File ".../google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/http_wrapper.py", line 356, in MakeRequest
    max_retry_wait, total_wait_sec))
  File ".../google-cloud-sdk/platform/gsutil/gslib/utils/retry_util.py", line 65, in WarnAfterManyRetriesHandler
    http_wrapper.HandleExceptionsAndRebuildHttpConnections(retry_args)
  File ".../google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/http_wrapper.py", line 304, in HandleExceptionsAndRebuildHttpConnections
    raise retry_args.exc
  File ".../google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/http_wrapper.py", line 346, in MakeRequest
    check_response_func=check_response_func)
  File ".../google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/http_wrapper.py", line 396, in _MakeRequestNoRetry
    redirections=redirections, connection_type=connection_type)
  File ".../google-cloud-sdk/platform/gsutil/gslib/vendored/oauth2client/oauth2client/transport.py", line 175, in new_request
    redirections, connection_type)
  File ".../google-cloud-sdk/platform/gsutil/gslib/vendored/oauth2client/oauth2client/transport.py", line 282, in request
    connection_type=connection_type)
  File ".../google-cloud-sdk/platform/gsutil/third_party/httplib2/python3/httplib2/__init__.py", line 1514, in request
    (response, content) = self._request(conn, authority, uri, request_uri, method, body, headers, redirections, cachekey)
  File ".../google-cloud-sdk/platform/gsutil/third_party/httplib2/python3/httplib2/__init__.py", line 1264, in _request
    (response, content) = self._conn_request(conn, request_uri, method, body, headers)
  File ".../google-cloud-sdk/platform/gsutil/third_party/httplib2/python3/httplib2/__init__.py", line 1187, in _conn_request
    conn.connect()
  File ".../google-cloud-sdk/platform/gsutil/third_party/httplib2/python3/httplib2/__init__.py", line 1011, in connect
    sock.connect((self.host, self.port))
  File ".../google-cloud-sdk/platform/gsutil/third_party/socksipy-branch/socks.py", line 383, in connect
    self.__negotiatehttp(destpair[0], destpair[1])
  File ".../google-cloud-sdk/platform/gsutil/third_party/socksipy-branch/socks.py", line 349, in __negotiatehttp
    raise HTTPError((statuscode, statusline[2]))
socks.HTTPError: (403, b'Forbidden')
```

Our proxy blocks the connection.
As NO_PROXY contains the googleapis.com domain, expected to not go through the proxy.

It works fine when unsetting the proxy environment variables.
Note that both HTTPS_PROXY and HTTP_PROXY need to be unset (the last one is surprising to me).

**Comment by turribeach** [comment]
Does it work if you set CLOUDSDK_GSUTIL_PYTHON to a Python 2 interpreter? I have seen issues with gsutil and Python 3.

**Comment by dilipped** [comment]
This should be fixed once https://github.com/GoogleCloudPlatform/gsutil/pull/1084 gets released. Will be released mostly by next week.

**Comment by dilipped** [comment]
As a workaround, you can try setting gsutil to use python2 as suggested by @turribeach.

**Comment by bquinart** [comment]
Looking forward to try out the released version.
I confirm it works with python2.
Thanks

**Comment by turribeach** [comment]
The issue I had has been fixed on gsutil v4.53 which is part of Cloud SDK 307.0.0. Worth giving yours a try as well on that version.

**Comment by bquinart** [comment]
I confirm this is fixed with v4.53. Thanks a lot!
